enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+to reconcile `IngressController` CRs due to this invalid combination.
+
+#### Support Procedures
+
+- For administrators and support engineers, the `IngressController` is still the source
+of truth and where you need to look for understanding the router placement in practice.
+Nothing has changed in that regard.
+
+## Implementation History
+
+Not yet applicable
+
+## Drawbacks
+
+- The pinning of the `IngressController` to the "master" pool is another change
+which would make single-node clusters slightly different from multi-node
+clusters, and any such difference is naturally not ideal.
+
+- The proposed defaulting behavior for the discussed `IngressController`
+parameters is complicated and dependent on three different parameters (infra
+topology, control-plane topology, and ingress placement) - such complexity
+would probably have to be documented in the CRD definitions and may confuse
+users.
+
+## Alternatives
+
+- Even when users need to add just one extra worker, require them to add yet
+another worker so they could just form a compact 3-node cluster where all
+nodes are both workers and control-plane nodes. This kind of topology is
+already supported by OCP. This will avoid the need for OCP to support yet
+another topology. It has the obvious downside of requiring a "useless" node
+the user didn't really need. It also means the user now has to run more
+control-plane workloads to facilitate HA - for example, 2 extra replicas of
+the API server which consume a lot of memory resources. From an engineering
+perspective, it would require us to make the "Control-plane Topology"
+parameter dynamic and make sure all operators know to react to changes in that
+parameter (it will have to change from `SingleReplica` to `HighlyAvailable`
+once those 2 new control-plane nodes join the cluster). I am not aware of the
+other engineering difficulties we would encounter when attempting to support
+the expansion of a single-node control-plane into a three-node control-plane,
+but I think they would not be trivial.
+
+- Adjust the "baremetal" platform to support single control-plane node
+installations and make users and the Assisted Installer (the current popular,
+supported method to install single control-plane node clusters) use that
+platform instead of the "none" platform. The baremetal platform solves the
+issue described in this enhancement with virtual IP addresses/keepalived. This
+approach was dismissed due to much higher development efforts and additional
+processes that would need to run on the already resource constrained single
+control-plane node. Furthermore, even once the baremetal platform is adjusted
+to support single-node clusters, the Assisted Installer would have to go
+through a lot of development effort in order to make it use the baremetal
+platform rather than the "none" platform currently used for single node
+installations. This may happen in the future.
+
```

Comment:
Another alternative would be to not change the installer at all, but to require users to set the IngressPlacement value before adding a worker.

Reply:
That would be too late, as the IngressController has already been created at that point with the wrong values. 

Also users could always just fix the IngressController manually (guess that could be listed as an alternative, although it's a bit of "don't enhance anything"), but the idea of this enhancement was to make the situation better with sane defaults so users don't encounter this ingress pitfall.

Reply:
I had a suggestion along these lines

> * Add both a `nodePlacement` field and a `replicas` field to the ingress config; the ingress operator could directly use these values, if provided, as default values for the corresponding fields in ingresscontrollers.  

https://github.com/openshift/enhancements/pull/996#discussion_r800730204

Maybe worth reconsidering?

Reply:
Maybe this thread isn't the best place for my comment.  It doesn't solve the fact that something (`ingressPlacement` or `nodePlacement`) needs to be set at install-time in the ingress config.  It's an alternative though to more complicated defaulting logic.

Reply:
If we want to go with `nodePlacement` and `replicas` instead of `ingressPlacement`, I can live with that. Either way, the installer has to make the same decision and have the same domain knowledge.

Reply:
The API alternative proposed here should be documented in the body of the enhancement. I think we say the reason we're not choosing that approach is that it would make the interface with the installer more complex, especially if the configuration API for Ingress changes in the future.

Reply:
Due to nodePlacement being a rather complex parameter, and it possibly having values other than the ones we're interested in (e.g. [the "linux" label](https://github.com/openshift/cluster-ingress-operator/blob/a6c16b7eefe6d7045f3109cd265c4e891f5791ef/pkg/operator/controller/ingress/deployment.go#L569)), I feel like the installation process dictating the exact values for these fields it's a bit outside of its expertise and a sign that maybe a more general concept like the proposed API is a better fit here

Reply:
Alternatives section updated to discuss this idea

Reply:
@omertuc I agree, but we should include that in the body of the enhancement so that future readers can see that it was considered.